### ISSUE TITLE
Boundary workers encounter errors connecting to controllers over http

### ISSUE THREAD

**Issue by Brcrwilliams** [hashicorp/boundary#847]
**Describe the bug**

I have a worker trying to connect to a boundary controller with TLS disabled. I am getting these error messages in the logs:

```
Jan 05 18:23:51 ip-10-0-0-9 systemd[1]: Started boundary worker.
Jan 05 18:23:51 ip-10-0-0-9 boundary[2397]: ==> Boundary server configuration:
Jan 05 18:23:51 ip-10-0-0-9 boundary[2397]:    AWS KMS KeyID: cc6a6013-883f-461f-9bff-7d7495ade606
Jan 05 18:23:51 ip-10-0-0-9 boundary[2397]:   AWS KMS Region: us-east-1
Jan 05 18:23:51 ip-10-0-0-9 boundary[2397]:              Cgo: disabled
Jan 05 18:23:51 ip-10-0-0-9 boundary[2397]:       Listener 1: tcp (addr: "0.0.0.0:9202", max_request_duration: "1m30s", purpose: "proxy")
Jan 05 18:23:51 ip-10-0-0-9 boundary[2397]:        Log Level: info
Jan 05 18:23:51 ip-10-0-0-9 boundary[2397]:            Mlock: supported: true, enabled: true
Jan 05 18:23:51 ip-10-0-0-9 boundary[2397]:      Public Addr: [redacted]:9202
Jan 05 18:23:51 ip-10-0-0-9 boundary[2397]:          Version: Boundary v0.1.4
Jan 05 18:23:51 ip-10-0-0-9 boundary[2397]:      Version Sha: f8577519b6fb152ddb19b2e4a7dcf8e9b1e82f58
Jan 05 18:23:51 ip-10-0-0-9 boundary[2397]: ==> Boundary server started! Log data will stream in below:
Jan 05 18:23:51 ip-10-0-0-9 boundary[2397]: 2021-01-05T18:23:51.617Z [INFO]  worker: connected to controller: address=10.0.0.11:9201
Jan 05 18:23:54 ip-10-0-0-9 boundary[2397]: 2021-01-05T18:23:54.118Z [ERROR] worker: error making status request to controller: error="rpc error: code = Unavailable desc = last connection error: connection error: desc = "transport: Error while dialing unable to write connection nonce: tls: server selected unsupported protocol version 303""
Jan 05 18:23:55 ip-10-0-0-9 boundary[2397]: 2021-01-05T18:23:55.838Z [ERROR] worker: error making status request to controller: error="rpc error: code = Unavailable desc = last connection error: connection error: desc = "transport: Error while dialing unable to write connection nonce: tls: server selected unsupported protocol version 303""
Jan 05 18:23:57 ip-10-0-0-9 boundary[2397]: 2021-01-05T18:23:57.767Z [ERROR] worker: error making status request to controller: error="rpc error: code = Unavailable desc = last connection error: connection error: desc = "transport: Error while dialing unable to write connection nonce: tls: server selected unsupported protocol version 303""
```

It _seems_ to me like the worker is trying to use TLS even though I have TLS disabled on the controller. I note that the [worker stanza](https://www.boundaryproject.io/docs/configuration/worker) has no way to configure if the worker should try to connect using TLS or not.

My worker configuration file is:

```hcl
listener "tcp" {
  address = "0.0.0.0:9202"
  purpose = "proxy"

  tls_disable   = true

}

worker {
  # Name attr must be unique
  public_addr = "[redacted]"
  name        = "demo-worker-0"
  description = "A default worker created for demonstration"
  controllers = [
    "10.0.0.11",
    "10.0.0.21",
  ]
}


kms "awskms" {
  purpose    = "worker-auth"
  key_id     = "global_root"
  kms_key_id = "cc6a6013-883f-461f-9bff-7d7495ade606"
}
```

and my controller configuration file is:

```hcl
disable_mlock = true

telemetry { 
  prometheus_retention_time = "24h"
  disable_hostname          = true
}

controller {
  name        = "demo-controller-1"
  description = "A controller for a demo!"

  public_cluster_addr = "[redacted]:9200"

  database {
    url = "[redacted]"
  }
}

listener "tcp" {
  address                           = "0.0.0.0:9200"
  purpose                           = "api"

  tls_disable                       = true

  cors_enabled                      = true
  cors_allowed_origins              = ["*"]
}

listener "tcp" {
  address = "0.0.0.0:9201"
  purpose = "cluster"

  tls_disable = true
}


kms "awskms" {
  purpose    = "root"
  key_id     = "global_root"
  kms_key_id = "e6b8341c-0e88-4015-8e24-d1bbf0896024"
}

kms "awskms" {
  purpose    = "worker-auth"
  key_id     = "global_worker_auth"
  kms_key_id = "cc6a6013-883f-461f-9bff-7d7495ade606"
}

kms "awskms" {
  purpose    = "recovery"
  key_id     = "global_recovery"
  kms_key_id = "d07ce89a-b1f3-4a97-bf9f-945057ad802c"
}
```

**To Reproduce**
1. Start boundary controller on a machine with the config above
2. Change controller IPs in the boundary worker config to your controller's IP
3. Start boundary worker with the config above
4. Boundary worker should produce TLS errors

**Expected behavior**
There should be some way to disable TLS on the worker -> controller connections

**Additional context**

**Comment by malnick** [comment]
Thanks for raising this @Brcrwilliams. The worker connects to the controller with TLS always, which is why it's required to have a kms stanza for the `worker-auth` purpose. You can read more about this [in our concept docs](https://www.boundaryproject.io/docs/concepts/security/connections-tls#worker-to-controller-tls). We currently do not support disabling TLS between the worker and controller. 

Can you show me the log output from the boundary controller when this error is happening on the worker side? I also would like to confirm that both the controller and worker are on hosts that have the necessary IAM permissions to use that KMS key?

**Comment by Brcrwilliams** [comment]
@malnick On the boundary controller I see this:

```
Jan 05 18:09:13 ip-10-0-0-21 systemd[1]: Started boundary controller.
Jan 05 18:09:13 ip-10-0-0-21 boundary[1832]: ==> Boundary server configuration:
Jan 05 18:09:13 ip-10-0-0-21 boundary[1832]:         AWS KMS KeyID: d07ce89a-b1f3-4a97-bf9f-945057ad802c
Jan 05 18:09:13 ip-10-0-0-21 boundary[1832]:         AWS KMS KeyID: d07ce89a-b1f3-4a97-bf9f-945057ad802c
Jan 05 18:09:13 ip-10-0-0-21 boundary[1832]:         AWS KMS KeyID: d07ce89a-b1f3-4a97-bf9f-945057ad802c
Jan 05 18:09:13 ip-10-0-0-21 boundary[1832]:        AWS KMS Region: us-east-1
Jan 05 18:09:13 ip-10-0-0-21 boundary[1832]:        AWS KMS Region: us-east-1
Jan 05 18:09:13 ip-10-0-0-21 boundary[1832]:        AWS KMS Region: us-east-1
Jan 05 18:09:13 ip-10-0-0-21 boundary[1832]:                   Cgo: disabled
Jan 05 18:09:13 ip-10-0-0-21 boundary[1832]:            Listener 1: tcp (addr: "0.0.0.0:9200", max_request_duration: "1m30s", purpose: "api")
Jan 05 18:09:13 ip-10-0-0-21 boundary[1832]:            Listener 2: tcp (addr: "0.0.0.0:9201", max_request_duration: "1m30s", purpose: "cluster")
Jan 05 18:09:13 ip-10-0-0-21 boundary[1832]:             Log Level: info
Jan 05 18:09:13 ip-10-0-0-21 boundary[1832]:                 Mlock: supported: true, enabled: false
Jan 05 18:09:13 ip-10-0-0-21 boundary[1832]:   Public Cluster Addr: boundary.garbage.engineering:9200
Jan 05 18:09:13 ip-10-0-0-21 boundary[1832]:               Version: Boundary v0.1.4
Jan 05 18:09:13 ip-10-0-0-21 boundary[1832]:           Version Sha: f8577519b6fb152ddb19b2e4a7dcf8e9b1e82f58
Jan 05 18:09:13 ip-10-0-0-21 boundary[1832]: ==> Boundary server started! Log data will stream in below:
Jan 05 18:23:56 ip-10-0-0-21 boundary[1832]: 2021-01-05T18:23:56.658Z [INFO]  controller: worker successfully authed: name=demo-worker-0
Jan 05 18:28:25 ip-10-0-0-21 boundary[1832]: 2021-01-05T18:28:25.343Z [INFO]  controller: worker successfully authed: name=demo-worker-0
```

I've verified that both the worker and the controller can use the KMS key by installing the AWS CLI on the instances and running it.

Worker: 

```
$ aws kms encrypt --key-id cc6a6013-883f-461f-9bff-7d7495ade606 --plaintext foo --region us-east-1
{
    "CiphertextBlob": "AQICAHjRkkIQ88zwlkVN2gIEyQL2S6H5H4MThnt1gEBtcSUaNQE24smSDeUL+nPkLUvVUe0FAAAAYTBfBgkqhkiG9w0BBwagUjBQAgEAMEsGCSqGSIb3DQEHATAeBglghkgBZQMEAS4wEQQMkfZodAkquE4gpdsMAgEQgB7CTsvbBVtS0tc7JsQKBlv+gZd8dg1JVCGi820iAcw=",
    "KeyId": "arn:aws:kms:us-east-1:[acct_id]:key/cc6a6013-883f-461f-9bff-7d7495ade606",
    "EncryptionAlgorithm": "SYMMETRIC_DEFAULT"
}
```

Controller:

```
$ aws kms decrypt --region us-east-1 --key-id cc6a6013-883f-461f-9bff-7d7495ade606 --ciphertext-blob fileb://ciphertext
{
    "KeyId": "arn:aws:kms:us-east-1:443424055065:key/cc6a6013-883f-461f-9bff-7d7495ade606",
    "Plaintext": "Zm9v",
    "EncryptionAlgorithm": "SYMMETRIC_DEFAULT"
}
$ echo 'Zm9v' | base64 -d
foo
```

This part of the error message `tls: server selected unsupported protocol version 303` seems to suggest that the controller listener is selecting TLS 1.2 as the TLS version (https://pkg.go.dev/crypto/tls#pkg-constants). Based on the documentation you linked, it should be using TLS 1.3. I tried changing my listener configuration on my controllers to:

```
listener "tcp" {
  address = "0.0.0.0:9201"
  purpose = "cluster"
  tls_min_version = "tls13"
}
```

and then restarted both the controllers and the worker, but I am still encountering the same error messaging.

**Comment by Brcrwilliams** [comment]
@malnick On the boundary controller I see this:

```
Jan 05 18:09:13 ip-10-0-0-21 systemd[1]: Started boundary controller.
Jan 05 18:09:13 ip-10-0-0-21 boundary[1832]: ==> Boundary server configuration:
Jan 05 18:09:13 ip-10-0-0-21 boundary[1832]:         AWS KMS KeyID: d07ce89a-b1f3-4a97-bf9f-945057ad802c
Jan 05 18:09:13 ip-10-0-0-21 boundary[1832]:         AWS KMS KeyID: d07ce89a-b1f3-4a97-bf9f-945057ad802c
Jan 05 18:09:13 ip-10-0-0-21 boundary[1832]:         AWS KMS KeyID: d07ce89a-b1f3-4a97-bf9f-945057ad802c
Jan 05 18:09:13 ip-10-0-0-21 boundary[1832]:        AWS KMS Region: us-east-1
Jan 05 18:09:13 ip-10-0-0-21 boundary[1832]:        AWS KMS Region: us-east-1
Jan 05 18:09:13 ip-10-0-0-21 boundary[1832]:        AWS KMS Region: us-east-1
Jan 05 18:09:13 ip-10-0-0-21 boundary[1832]:                   Cgo: disabled
Jan 05 18:09:13 ip-10-0-0-21 boundary[1832]:            Listener 1: tcp (addr: "0.0.0.0:9200", max_request_duration: "1m30s", purpose: "api")
Jan 05 18:09:13 ip-10-0-0-21 boundary[1832]:            Listener 2: tcp (addr: "0.0.0.0:9201", max_request_duration: "1m30s", purpose: "cluster")
Jan 05 18:09:13 ip-10-0-0-21 boundary[1832]:             Log Level: info
Jan 05 18:09:13 ip-10-0-0-21 boundary[1832]:                 Mlock: supported: true, enabled: false
Jan 05 18:09:13 ip-10-0-0-21 boundary[1832]:   Public Cluster Addr: [redacted]:9200
Jan 05 18:09:13 ip-10-0-0-21 boundary[1832]:               Version: Boundary v0.1.4
Jan 05 18:09:13 ip-10-0-0-21 boundary[1832]:           Version Sha: f8577519b6fb152ddb19b2e4a7dcf8e9b1e82f58
Jan 05 18:09:13 ip-10-0-0-21 boundary[1832]: ==> Boundary server started! Log data will stream in below:
Jan 05 18:23:56 ip-10-0-0-21 boundary[1832]: 2021-01-05T18:23:56.658Z [INFO]  controller: worker successfully authed: name=demo-worker-0
Jan 05 18:28:25 ip-10-0-0-21 boundary[1832]: 2021-01-05T18:28:25.343Z [INFO]  controller: worker successfully authed: name=demo-worker-0
```

I've verified that both the worker and the controller can use the KMS key by installing the AWS CLI on the instances and running it.

Worker: 

```
$ aws kms encrypt --key-id cc6a6013-883f-461f-9bff-7d7495ade606 --plaintext foo --region us-east-1
{
    "CiphertextBlob": "AQICAHjRkkIQ88zwlkVN2gIEyQL2S6H5H4MThnt1gEBtcSUaNQE24smSDeUL+nPkLUvVUe0FAAAAYTBfBgkqhkiG9w0BBwagUjBQAgEAMEsGCSqGSIb3DQEHATAeBglghkgBZQMEAS4wEQQMkfZodAkquE4gpdsMAgEQgB7CTsvbBVtS0tc7JsQKBlv+gZd8dg1JVCGi820iAcw=",
    "KeyId": "arn:aws:kms:us-east-1:[acct_id]:key/cc6a6013-883f-461f-9bff-7d7495ade606",
    "EncryptionAlgorithm": "SYMMETRIC_DEFAULT"
}
```

Controller:

```
$ aws kms decrypt --region us-east-1 --key-id cc6a6013-883f-461f-9bff-7d7495ade606 --ciphertext-blob fileb://ciphertext
{
    "KeyId": "arn:aws:kms:us-east-1:[acct_id]:key/cc6a6013-883f-461f-9bff-7d7495ade606",
    "Plaintext": "Zm9v",
    "EncryptionAlgorithm": "SYMMETRIC_DEFAULT"
}
$ echo 'Zm9v' | base64 -d
foo
```

This part of the error message `tls: server selected unsupported protocol version 303` seems to suggest that the controller listener is selecting TLS 1.2 as the TLS version (https://pkg.go.dev/crypto/tls#pkg-constants). Based on the documentation you linked, it should be using TLS 1.3. I tried changing my listener configuration on my controllers to:

```
listener "tcp" {
  address = "0.0.0.0:9201"
  purpose = "cluster"
  tls_min_version = "tls13"
}
```

and then restarted both the controllers and the worker, but I am still encountering the same error messaging.

**Comment by malnick** [comment]
Roping in @jefferai to repro and look into this one, thanks for the additional info @Brcrwilliams

**Comment by jefferai** [comment]
It looks like you're sending the traffic from the worker through some proxy or load balancer that is trying to decrypt TLS. We don't support this; we carefully manage TLS between the controller and worker, and part of that is that they both only support TLS 1.3. That "version 303" in the error message means the "server" is announcing a maximum TLS version of TLS 1.2 -- which suggests that it's not actually talking to the Boundary controller.

You're seeing successful authentication because that happens as a part of the TLS handshake, but _before_ maximum TLS versions are negotiated (that is, the authentication happens, then returns the TLS configuration that should be used).

**Comment by jefferai** [comment]
Oh, likely related:

`public_cluster_addr = "[redacted]:9200"`

This should probably be port 9201. Otherwise what will happen is the workers will connect, then the controller will give them updated addresses -- which have the wrong port. That would cause a TLS mismatch, certainly.

**Comment by Brcrwilliams** [comment]
@jefferai That was it! I completely misunderstood the purpose of `public_cluster_addr` and I set it to my load balancer hostname.  I changed it to the instance ip address instead and that solved the issue.

I think there is a small mistake on the documentation: https://www.boundaryproject.io/docs/configuration/controller#public_cluster_addr

> Specifies the public host or IP address (and optionally port) at which the worker can be reached by workers. 

I think this should say "[the address] at which the _controller_ can be reached by workers" ?
